### ISSUE TITLE
Postfix causes 100% CPU and is crashing

### ISSUE THREAD

**Issue by thannaske** [mailcow/mailcow-dockerized#1988]
**Describe the bug**
The mailcow stack is unexpectedly producing a 100% CPU load on the system. I suspect that the Postfix-part is responsible for this problems as it refuses any connection during the problems. Furthermore I observed that Postfix seems to crash from time to time. I restarted the whole system when I detected the problems with the CPU load. Afterwards the stack came back online and Postfix immediately crashes after a few seconds/minutes uptime (see log lines attached).  

**How to reproduce Reproduce**
A step-by-step reproduction is currently not possible as the behavior became visible unexpectedly.

**Expected behavior**
Postfix shall accept connections, shall not crash and shall not produce 100% load.

**Screenshots**
<img width="1169" alt="bildschirmfoto 2018-11-06 um 10 43 20" src="https://user-images.githubusercontent.com/8004696/48056321-cd9ed680-e1b1-11e8-9816-121b46635c39.png">

**Logs**
```
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:24:54 vsc03 postfix/qmgr[356]: A4A4446AF3: removed
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:19 vsc03 postfix/postscreen[429]: CONNECT from [185.38.119.204]:45542 to [172.22.1.8]:25
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:19 vsc03 whitelist_forwardinghosts: Look up 185.38.119.204 on whitelist, result 200 DUNNO
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:19 vsc03 postfix/dnsblog[442]: addr 185.38.119.204 listed by domain hostkarma.junkemailfilter.com as 127.0.0.3
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:19 vsc03 postfix/dnsblog[442]: addr 185.38.119.204 listed by domain hostkarma.junkemailfilter.com as 127.0.1.1
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:22 vsc03 postfix/postscreen[429]: PASS NEW [185.38.119.204]:45542
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:29 vsc03 postfix/dnsblog[439]: warning: dnsblog_query: lookup error for DNS query 204.119.38.185.zen.spamhaus.org: Host or domain name not found. Name service error for name=204.119.38.185.zen.spamhaus.org type=A: Host not found, try again
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:29 vsc03 postfix/postscreen[429]: fatal: watchdog timeout
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:30 vsc03 postfix/master[354]: warning: process /usr/lib/postfix/sbin/postscreen pid 429 exit status 1
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:44,890 WARN received SIGTERM indicating exit request
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:44,890 INFO waiting for postfix, processes, syslog-ng to die
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:25:44 vsc03 syslog-ng[10]: syslog-ng shutting down; version='3.13.2'
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,003 INFO stopped: syslog-ng (exit status 0)
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,040 INFO reaped unknown pid 354
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,040 INFO reaped unknown pid 357
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,041 INFO reaped unknown pid 365
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,041 INFO reaped unknown pid 371
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,041 INFO reaped unknown pid 359
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 373
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 427
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 439
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 440
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 441
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,042 INFO reaped unknown pid 442
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,043 INFO reaped unknown pid 443
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,043 INFO reaped unknown pid 444
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,044 INFO reaped unknown pid 391
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,045 INFO reaped unknown pid 364
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,046 INFO reaped unknown pid 355
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,046 INFO reaped unknown pid 356
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,046 INFO reaped unknown pid 368
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,047 INFO reaped unknown pid 372
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,047 INFO reaped unknown pid 428
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,047 INFO reaped unknown pid 437
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,048 INFO reaped unknown pid 438
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,051 INFO stopped: postfix (terminated by SIGTERM)
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:25:45,052 INFO stopped: processes (terminated by SIGTERM)
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:28:18,128 CRIT Set uid to user 0
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:28:18,132 INFO supervisord started with pid 1
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:28:19,135 INFO spawned: 'processes' with pid 10
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:28:19,138 INFO spawned: 'postfix' with pid 11
postfix-mailcow_1_8cf49c64f461 | 2018-11-06 10:28:19,141 INFO spawned: 'syslog-ng' with pid 12
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 syslog-ng[12]: syslog-ng starting up; version='3.13.2'
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 postfix[270]: Postfix is running with backwards-compatible default settings
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 postfix[270]: See http://www.postfix.org/COMPATIBILITY_README.html for details
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 postfix[270]: To disable backwards compatibility use "postconf compatibility_level=2" and "postfix reload"
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 postfix/postfix-script[355]: starting the Postfix mail system
postfix-mailcow_1_8cf49c64f461 | Nov  6 10:28:19 vsc03 postfix/master[357]: daemon started -- version 3.3.0, configuration /opt/postfix/conf
```

**System**
```
root@vsc03 /opt/mailcow-dockerized # docker --version
Docker version 18.06.1-ce, build e68fc7a

root@vsc03 /opt/mailcow-dockerized # docker-compose --version
docker-compose version 1.23.1, build b02f1306

root@vsc03 /opt/mailcow-dockerized # cat /etc/debian_version 
9.5
```
mailcow-dockerized is running on the latest version. I literally updated minutes ago.

**Comment by andryyy** [comment]
Hi,

Unfortunately we need more information. Like a "ps aux" when it happens, to find out which process is stuck.

**Comment by thannaske** [comment]
Will try to do so when it crashes again. Currently it's running stable for the last 30 minutes.

**Comment by andryyy** [comment]
Am eager to see what's causing this. :-)

**Comment by chriscroome** [comment]
I'm seeing the same thing happening a few time each night and could post similar log output as above.

Last night it happened at 1:10am and 6:29am and the day before at 1:02am, 6:27am and 6:32am and the night before at 2:04am and 6:28am &mdash; perhaps these are times where a lot of the (other) servers are running backups or regular tasks and therefore the disk IO is worse, or perhaps it is related to an issue in Mailcow or a combination of both? It only started happening after updating Mailcow on 28th October 2018.

**Comment by andryyy** [comment]
Please post your logs, but no monitoring graphs, they don't help to find the cause. :-)

Does it crash your server or at least postfix-mailcow?

Please try to monitor the processes of the node.

**Comment by chriscroome** [comment]
OK, the last shutdown and restart:

```
postfix-mailcow_1    | 2018-11-06 06:29:16,186 WARN received SIGTERM indicating exit request
postfix-mailcow_1    | 2018-11-06 06:29:16,186 INFO waiting for postfix, syslog-ng to die
postfix-mailcow_1    | Nov  6 06:29:16 mail syslog-ng[11]: syslog-ng shutting down; version='3.13.2'
postfix-mailcow_1    | 2018-11-06 06:29:16,292 INFO stopped: syslog-ng (exit status 0)
postfix-mailcow_1    | 2018-11-06 06:29:16,356 INFO reaped unknown pid 382
postfix-mailcow_1    | 2018-11-06 06:29:16,356 INFO reaped unknown pid 385
postfix-mailcow_1    | 2018-11-06 06:29:16,357 INFO reaped unknown pid 390
postfix-mailcow_1    | 2018-11-06 06:29:16,358 INFO reaped unknown pid 384
postfix-mailcow_1    | 2018-11-06 06:29:16,358 INFO reaped unknown pid 396
postfix-mailcow_1    | 2018-11-06 06:29:16,358 INFO reaped unknown pid 397
postfix-mailcow_1    | 2018-11-06 06:29:16,359 INFO reaped unknown pid 6336
postfix-mailcow_1    | 2018-11-06 06:29:16,359 INFO reaped unknown pid 6354
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6401
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6445
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6489
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6490
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6507
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6524
postfix-mailcow_1    | 2018-11-06 06:29:16,360 INFO reaped unknown pid 6525
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6526
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6527
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6528
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6529
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6530
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6550
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6551
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6552
postfix-mailcow_1    | 2018-11-06 06:29:16,361 INFO reaped unknown pid 6562
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6590
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6605
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6607
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6608
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6610
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6613
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6614
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6616
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6619
postfix-mailcow_1    | 2018-11-06 06:29:16,362 INFO reaped unknown pid 6620
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6622
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO stopped: postfix (terminated by SIGTERM)
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6624
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6625
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6626
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6627
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6655
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6706
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6731
postfix-mailcow_1    | 2018-11-06 06:29:16,363 INFO reaped unknown pid 6751
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6752
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6753
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6756
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6758
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6767
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6771
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6782
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6783
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6792
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6801
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6805
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6806
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6816
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6817
postfix-mailcow_1    | 2018-11-06 06:29:16,364 INFO reaped unknown pid 6834
postfix-mailcow_1    | 2018-11-06 06:29:16,365 INFO reaped unknown pid 6836
postfix-mailcow_1    | 2018-11-06 06:29:16,365 INFO reaped unknown pid 6837
postfix-mailcow_1    | 2018-11-06 06:29:16,365 INFO reaped unknown pid 6759
postfix-mailcow_1    | 2018-11-06 06:29:16,365 INFO reaped unknown pid 6855
postfix-mailcow_1    | 2018-11-06 06:29:40,283 CRIT Set uid to user 0
postfix-mailcow_1    | 2018-11-06 06:29:40,287 INFO supervisord started with pid 1
postfix-mailcow_1    | 2018-11-06 06:29:41,297 INFO spawned: 'postfix' with pid 10
postfix-mailcow_1    | 2018-11-06 06:29:41,305 INFO spawned: 'syslog-ng' with pid 11
postfix-mailcow_1    | Nov  6 06:29:41 mail syslog-ng[11]: syslog-ng starting up; version='3.13.2'
```

Before this are the usual logs lines (which I could post but would take some time to make anonymous), plus:

```
postfix-mailcow_1    | Nov  6 06:27:16 mail postfix/postscreen[6669]: fatal: watchdog timeout
...
postfix-mailcow_1    | Nov  6 06:28:15 mail postfix/postscreen[6698]: fatal: watchdog timeout

```

**Comment by andryyy** [comment]
Everything before "WARN received SIGTERM indicating exit request" would be useful, something sent a SIGTERM. It was probably sent by watchdog-mailcow anyway.

Could be related to whitelist_forwardinghosts.sh which gets stuck. Anything in the logs regarding "whitelist_forwardinghosts.sh"? Or "command time limit exceeded"?

We could try to add a timeout to the curl command in whitelist_forwardinghosts.sh. Vsevolod will move the RBL checks to a pre-data stage of the milter check so we can move away from these checks in Postscreen anyway.

Any useful logs in Redis?

**Comment by thannaske** [comment]
> perhaps these are times where a lot of the (other) servers are running backups or regular tasks and therefore the disk IO is worse, or perhaps it is related to an issue in Mailcow or a combination of both?

Interesting. My backup is running at 07:57 in the morning. The issues however started 20 minutes earlier so I guess this is not related to this process. In addition to the high CPU I could observe a huge increase in memory usage (up to 95%, guess it started to swap as I can see a huge I/O load when the RAM hit that level).

I will continue to monitor this issue; it did not happen again since this morning. But "good" to hear that I'm not alone with that problem. Those kind of issues never happened to me before performing the 28th October update, so I would carefully guess that it could be related to this.

> Could be related to whitelist_forwardinghosts.sh which gets stuck. Anything in the logs regarding "whitelist_forwardinghosts.sh"? Or "command time limit exceeded"?

Redis were flushed upon restart so I can't deliver any logs regarding the timestamp between the start of the issues and my first restart. After that restart the CPU load remained high for another 15 minutes but I could not find any log lines that contained the given strings. There were just a few lines including `whitelist_forwardinghosts: Look up $ip on whitelist, result 200 DUNNO` but that seems normal to me, isn't it?

**Comment by andryyy** [comment]
That's indeed okay. :-)

Would be nice if you guys monitor the procs for a while. Atop is great for this.

**Comment by thannaske** [comment]
Feedback for you: It did not happen again - at least not in this way. 

I experienced another crash a few days later that showed nearly the same monitoring footprint. But thanks to atop I was able to correlate the crash to another non-mailcow-related process that consumed so much memory that the server ran into the swap. It's possible that this also was the reason for the failure that lead to this issue. So I guess we can close this at this point.
